src/test/java/com/commercetools/sync/customers/helpers/CustomerReferenceResolverTest.java

```diff
@@ -0,0 +1,352 @@
+
+        // Test and assertion
+        final String expectedExceptionMessage =
+            format(FAILED_TO_RESOLVE_CUSTOM_TYPE, customerDraftBuilder.getKey());
+        final String expectedMessageWithCause =
+            format("%s Reason: %s", expectedExceptionMessage, format(TYPE_DOES_NOT_EXIST, customTypeKey));
+        assertThat(referenceResolver.resolveCustomTypeReference(customerDraftBuilder))
+            .hasFailedWithThrowableThat()
+            .isExactlyInstanceOf(ReferenceResolutionException.class)
+            .hasMessage(expectedMessageWithCause);
+    }
+
+    @Test
+    void resolveReferences_WithAllValidReferences_ShouldResolveReferences() {
+        final CustomerDraft customerDraft = CustomerDraftBuilder
+            .of("email@example.com", "secret123")
+            .key("customer-key")
+            .custom(CustomFieldsDraft.ofTypeKeyAndJson("typeKey", emptyMap()))
+            .build();
+
+        final CustomerDraft resolvedDraft = referenceResolver
+            .resolveReferences(customerDraft)
+            .toCompletableFuture()
+            .join();
+
+        final CustomerDraft expectedDraft = CustomerDraftBuilder
+            .of(customerDraft)
+            .custom(CustomFieldsDraft.ofTypeIdAndJson("typeId", new HashMap<>()))
+            .build();
+
+        assertThat(resolvedDraft).isEqualTo(expectedDraft);
+    }
+
+    @Test
+    void resolveCustomerGroupReference_WithKeys_ShouldResolveReference() {
+        final CustomerDraftBuilder customerDraftBuilder = CustomerDraftBuilder
+            .of("email@example.com", "secret123")
+            .customerGroup(ResourceIdentifier.ofKey("customer-group-key"));
+
+        final CustomerDraftBuilder resolvedDraft = referenceResolver.resolveCustomerGroupReference(customerDraftBuilder)
+                                                                    .toCompletableFuture().join();
+
+        assertThat(resolvedDraft.getCustomerGroup()).isNotNull();
+        assertThat(resolvedDraft.getCustomerGroup().getId()).isEqualTo(CUSTOMER_GROUP_ID);
+    }
+
+    @Test
+    void resolveCustomerGroupReference_WithNonExistentCustomerGroup_ShouldNotResolveReference() {
+        final CustomerDraftBuilder customerDraftBuilder = CustomerDraftBuilder
+            .of("email@example.com", "secret123")
+            .customerGroup(ResourceIdentifier.ofKey("anyKey"))
+            .key("dummyKey");
+
+        when(customerGroupService.fetchCachedCustomerGroupId(anyString()))
+            .thenReturn(completedFuture(Optional.empty()));
+
+        final String expectedMessageWithCause = format(FAILED_TO_RESOLVE_CUSTOMER_GROUP_REFERENCE,
+            "dummyKey", format(CUSTOMER_GROUP_DOES_NOT_EXIST, "anyKey"));
+
+        referenceResolver.resolveCustomerGroupReference(customerDraftBuilder)
+                         .exceptionally(exception -> {
+                             assertThat(exception).hasCauseExactlyInstanceOf(ReferenceResolutionException.class);
```

Comment:
If the test doesn't go to `.exceptionally` block, this test will also pass?

Reply:
Need to check, probably a hasFailed check is required.

Reply:
328f92e